wdl-analysis/src/handlers/find_all_references.rs

```diff
@@ -0,0 +1,227 @@
+//! Implements find all references functionality
+
+use std::sync::Arc;
+
+use anyhow::Context;
+use anyhow::Result;
+use anyhow::anyhow;
+use line_index::LineIndex;
+use lsp_types::Location;
+use url::Url;
+use wdl_ast::AstNode;
+use wdl_ast::SyntaxKind;
+use wdl_ast::TreeToken;
+
+use super::goto_definition::position_to_offset;
+use crate::Document;
+use crate::SourcePosition;
+use crate::SourcePositionEncoding;
+use crate::graph::DocumentGraph;
+use crate::graph::DocumentGraphExt;
+use crate::handlers;
+use crate::handlers::location_from_span;
+use crate::handlers::position;
+
+/// Represents a target definition for which references are being searched.
+#[derive(Debug)]
+struct TargetDefinition {
+    /// The identifier text of the target symbol.
+    name: String,
+    /// The syntax kind of the target's parent node, used to determine reference
+    /// context.
+    kind: SyntaxKind,
+    /// The location where the target is defined.
+    location: Location,
+}
+
+/// Finds all references to the identifier at the given position.
+///
+/// It first resolves the definition of the identifier at the specified
+/// position, then searches through the appropriate scope of
+/// documents to find all references to that definition.
+pub fn find_all_references(
+    graph: &DocumentGraph,
+    document_uri: Url,
+    position: SourcePosition,
+    encoding: SourcePositionEncoding,
+) -> Result<Vec<Location>> {
+    let definition_location = handlers::goto_definition(graph, document_uri, position, encoding)
+        .context("failed to resolve symbol definition")?
+        .ok_or_else(|| {
+            anyhow!(
+                "no definition location found for symbol at position: {}:{}",
+                position.line,
+                position.character
+            )
+        })?;
+
+    let target = resolve_target_definition(graph, &definition_location, encoding)
+        .context("failed to resolve target definition")?;
+
+    let search_scope = determine_search_scope(&target, graph);
+
+    let mut locations = Vec::new();
+    for doc_index in search_scope {
+        collect_references_from_document(graph, doc_index, &target, encoding, &mut locations)
+            .with_context(|| {
+                format!("failed to collect references from document at index {doc_index:?}")
+            })?;
+    }
+
+    Ok(locations)
+}
+
+/// Resolves the target definition from a definition location.
```

Comment:
I don't know what this means just based on this sentence alone. Can we refine this docstring?